### PR TITLE
Refactoring signup action

### PR DESCRIPTION
**Pull request by weerd**
### What does this PR do?

This PR helps establish a new, simpler flow for the signup action process. It helps simplify a few of the actions in the `actions/signup.js` file and makes things a bit easier to follow.

It stubs some code out in the `CampaignSignupsController` and `SignupRepository` for upcoming PR to connect sending signups to Rogue v3 endpoint.

@DFurnes and I also worked together to address using a simpler approach for handling post-authentication signups on a campaign. It now uses a new browser storage package called `localforage` that more robustly stores data in the browser, and combining this with Laravel's sessions to know when a user was redirected back to a campaign after logging in via DS Identity (Northstar) and picking back up the rest of the signup process.

### Any background context you want to provide?

We are aiming to get rid of the `actions/events.js`, `reducers/events.js` and associated `queue` in the redux store which currently is a bit convoluted. We're also going to get rid of the `helpers/storage.js` in favor of the `localforage` package which handles storing data in the browser and provides lots of useful methods.

### What are the relevant tickets/cards?

Refs [Pivotal ID #160602918](https://www.pivotaltracker.com/story/show/160602918)
Refs [Pivotal ID #152003980](https://www.pivotaltracker.com/story/show/152003980)

### CODE REVIEW

File: resources/assets/actions/signup.js [hunk 1]
```diff
@@ -111,10 +117,65 @@ export function getCampaignSignups(id = null, query = {}) {
...
-export function storeCampaignSignup() {}
+/**
+ * Dispatch action to store signup for a campaign.
+ *
+ * @param  {String} campaignId
+ * @param  {Object} data
+ * @return {Function}
+ */
+export function storeCampaignSignup(campaignId, data) {
+  // @TODO: if no campaignId or data let's throw an error!
```
Comment (DFurnes) [inline resources/assets/actions/signup.js]: Static typpppiiiiiing! 🧟‍♂️

Comment (weerd) [reply resources/assets/actions/signup.js]: via flow-types right? We should likely revisit post-Ashes migration 🤔

Comment (DFurnes) [reply resources/assets/actions/signup.js]: Yeah, Flow or Typescript would be the two front-runners here.

File: resources/assets/actions/signup.js [hunk 2]
```diff
@@ -111,10 +117,65 @@ export function getCampaignSignups(id = null, query = {}) {
...
+  };
+
+  return dispatch => {
+    dispatch(
+      apiRequest('POST', {
+        body: { ...data },
+        failure: 'failure_message',
+        meta: {
+          analytics,
+          sixpackExperiments: { conversion: 'signup' },
```
Comment (mendelB) [inline resources/assets/actions/signup.js]: Would this be causing a second conversion on the `'signup'` in addition to the dispatched `convertOnSignup` action in the `clickedSignupAction`?
https://github.com/DoSomething/phoenix-next/pull/1174/files#diff-9e724eddc3f7ebfa3cb156868580310aR170

File: resources/assets/actions/signup.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { Phoenix } from '@dosomething/gateway';
 import apiRequest from './api';
 import { isCampaignClosed } from '../helpers';
 import { getUserId, isAuthenticated } from '../selectors/user';
+import { getCampaignAffiliateMessagingOptOutFlag } from '../selectors/campaign';
```
Comment (mendelB) [inline resources/assets/actions/signup.js]: 🥇

File: app/Http/Controllers/AuthController.php [hunk 4]
```diff
@@ -46,10 +46,14 @@ public function getLogin(ServerRequestInterface $request, ResponseInterface $res
         if (! array_has($request->getQueryParams(), 'code')) {
             $intended = session()->pull('url.intended', url()->previous());
             session(['login.intended' => $intended]);
+
+            // If starting the login process, see if we have a "queued" action.
+            $actionId = array_get($request->getQueryParams(), 'actionId');
+            session()->flash('actionId', $actionId);
```
Comment (mendelB) [inline app/Http/Controllers/AuthController.php]: Why do we need to store the `actionId` in the session -- would we ever be redirected back to this action after having already hit it in a previous request?
(I also think I just don't have a good grasp on this Controller action, to begin with, which is why I may be misunderstanding something here.)

File: resources/assets/actions/signup.js (hunk 2 above)
Comment (weerd) [reply resources/assets/actions/signup.js]: Yeah, I was curious about that as well, but I don't think so. Sixpack only converts once for a user, regardless how many times a conversion happens, so I _think_ we're ok!

File: resources/assets/actions/signup.js (hunk 3 above)
Comment (weerd) [reply resources/assets/actions/signup.js]: Naming..., amiright 😉

File: resources/assets/actions/signup.js (hunk 2 above)
Comment (mendelB) [reply resources/assets/actions/signup.js]: Ah sweet!

File: app/Http/Controllers/AuthController.php (hunk 4 above)
Comment (weerd) [reply app/Http/Controllers/AuthController.php]: Oh ya. That is a good point. I believe it's still useful, but may not be 100% necessary. I'll confirm in follow up PR where I wrap up connecting this all with Rogue! But defs good question :)

**Pull request merged**